src/Hedgehog/Plutus/Tx.hs

```diff
@@ -0,0 +1,600 @@
+
+data TxIn = TxIn
+  { txInRef :: !Plutus.TxOutRef
+  , txInScript :: !(Maybe InScript)
+  }
+  deriving stock (Eq, Ord)
+
+data TxOut = TxOut
+  { txOutAddress :: !Plutus.Address
+  , txOutValue :: !Plutus.Value
+  , txOutDatum :: !(Maybe Plutus.Datum)
+  , txOutReferenceScript :: !(Maybe Script)
+  }
+
+newtype Script = Script
+  { unScript :: UPLC.Program UPLC.DeBruijn PLC.DefaultUni PLC.DefaultFun ()
+  }
+
+data InScript = InScript
+  { inScriptSource :: ScriptSource
+  , inScriptData :: Maybe (Plutus.Redeemer, Plutus.Datum)
+  }
+  deriving stock (Eq, Ord)
+
+data ScriptSource = InTransaction | RefScript !Plutus.TxOutRef
+  deriving stock (Eq, Ord)
+
+instance Semigroup (Tx bal) where
+  l <> r =
+    Tx
+      { txInputs = txInputs l <> txInputs r
+      , txOutputs = txOutputs l <> txOutputs r
+      , txMint = txMint l <> txMint r
+      , txFee = txFee l + txFee r
+      , txValidRange = txValidRange l /\ txValidRange r
+      , txExtraSignatures = txExtraSignatures l <> txExtraSignatures r
+      }
+
+instance Monoid (Tx bal) where
+  mempty =
+    Tx
+      { txInputs = Set.empty
+      , txOutputs = Vector.empty
+      , txMint = Map.empty
+      , txFee = 0
+      , txValidRange = Plutus.never
+      , txExtraSignatures = Set.empty
+      }
+
+data TxContext = TxContext
+  { mockchain :: !Model.Mock
+  , interestingScripts :: !(Map Plutus.ScriptHash Script)
+  , datums :: !(Map Plutus.DatumHash Plutus.Datum)
+  }
+
+data ScriptPurpose
+  = Minting
+      Plutus.CurrencySymbol
+      (Map Plutus.TokenName Integer, Plutus.Redeemer)
+  | Spending Plutus.TxOutRef InScript
+
+{- | Balance a transaction. Algorithm:
+
+ * Find the smallest set of 'non-special' UTxOs (PubKey locked no datum) that
+   cover the transaction balance.
+
+ * Pay change, if any, to a random PubKeyHash.
+
+  In the future, hps should support custom balancers.
+-}
+balanceTx :: TxContext -> Tx 'Unbalanced -> Maybe (Hedgehog.Gen (Tx 'Balanced))
+balanceTx context tx = balanceTxWhere context tx (const True) (const True)
+
+-- | as balanceTx but only uses txOuts belonging to a particular PubKeyHash and always sends change to that PubKeyHash
+balanceTxAsPubKey :: TxContext -> Tx 'Unbalanced -> PubKeyHash -> Maybe (Hedgehog.Gen (Tx 'Balanced))
+balanceTxAsPubKey context tx pkh =
+  balanceTxWhere
+    context
+    tx
+    ( Plutus.txOutAddress
+        >>> Plutus.addressCredential
+        >>> ( \case
+                Plutus.PubKeyCredential pkh' -> pkh' == pkh
+                _ -> False
+            )
+    )
+    (== pkh)
+
+-- | as balanceTx but with additional predicates about which TxOuts and PubKeyHashs can be used
+balanceTxWhere ::
+  TxContext ->
+  Tx 'Unbalanced ->
+  (Plutus.TxOut -> Bool) ->
+  (Plutus.PubKeyHash -> Bool) ->
+  Maybe (Hedgehog.Gen (Tx 'Balanced))
+balanceTxWhere context tx predTxout predPkh = do
+  let mock = mockchain context
+  (valueIn, valueOut) <- getValueInAndOut context tx
+  -- TODO we may want to change this to something like
```

Comment:
I've been pretty lassiez-faire about using `fromJust` for tx lookup. `Tx`s are only supposed to be valid under a particular `TxContext`, and you should only be passing in valid refs. In theory, we could use ghosts of departed proofs to enforce that, but that's a long way off...